Multiple transport interface

DanielYEHsieh opened the pull request:
<!--- Title -->

Description
-----------
Copying description from https://github.com/FreeRTOS/FreeRTOS/pull/434/files

By removing the definition of the NetworkContext struct in the header file, we allow the application to define it. This allows an application writer to use multiple transports in the same compilation unit. That way, multiple .c files do not have to be created for each transport.

Testing
-----------
PC:
demos
https://amazon-freertos-ci.corp.amazon.com/job/im_v2_master/job/nightly_custom_job/2000/
tests
https://amazon-freertos-ci.corp.amazon.com/job/im_v2_master/job/nightly_custom_job/1995/

All Passed

esp:
demos:
https://amazon-freertos-ci.corp.amazon.com/job/im_v2_master/job/nightly_custom_job/1999/
tests:
https://amazon-freertos-ci.corp.amazon.com/job/im_v2_master/job/nightly_custom_job/1996/#showFailuresLink

only BLE failed due to bcn timeout, but seems not related to the changes.
( tested without the changes still gets bcn timeout )

BLE demos passed locally(CI is not working, tested with esp32 locally):
CONFIG_SHADOW_BLE_TRANSPORT_DEMO_ENABLED passed
CONFIG_MQTT_BLE_TRANSPORT_DEMO_ENABLED pass
https://paste.amazon.com/show/dhs/1607975879
https://paste.amazon.com/show/dhs/1607975352



By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Commit:
support multiple transport interface co-existence

Commit:
run uncrustify

Commit:
transport_secure_sockets_utest.c 100%

Commit:
fix build failure for legacy MQTT path

Commit:
fix test project failuers

Commit:
fix for shadow demo

Commit:
fix Connect_LWT fail due to second networkContext not assign pParams.

Commit:
update demos transport params declaration

Commit:
fix typo

Commit:
fix demo failures

Commit:
move BLE context assign after memory is reset

leegeth commented:
/bot run checks

yourslab commented on libraries/c_sdk/standard/mqtt/CMakeLists.txt:
```diff
@@ -3,7 +3,13 @@ afr_module(INTERNAL)
+if( EXISTS ${AFR_MODULES_DIR}/coreMQTT/source )
+    set(transport_interface_dir "${AFR_MODULES_DIR}/coreMQTT/source/interface")
+elseif( EXISTS ${AFR_MODULES_DIR}/coreHTTP/source )
+    set(transport_interface_dir "${AFR_MODULES_DIR}/coreHTTP/source/interface")
```
How come we need to check coreHTTP too? I think coreMQTT is already SSoT for transport interface plus this build is for MQTT.

DanielYEHsieh replied:
Thanks for pointing that out, fixed in 6e29b18539d2f1145874f40d9661a458090c93b1

yourslab commented on libraries/c_sdk/standard/mqtt/src/iot_mqtt_api.c:
```diff
@@ -64,12 +64,30 @@
+ * @brief Represents the network context used for the TLS session with the
+ * server.
+ */
+static NetworkContext_t networkContext = { 0 };
```
This is static now. Could this cause problems? Can we possibly have a race condition with `pParams` when two tasks simultaneously make a connection?

DanielYEHsieh replied:
Thanks for the suggestion, fixed ini c1b87014da86c6e5d22bc60ad0efe824c5ef873c

yourslab commented on libraries/c_sdk/standard/mqtt/src/private/iot_mqtt_internal.h:
```diff
@@ -450,10 +453,10 @@ typedef struct connContextMapping
     StaticSemaphore_t contextMutexStorage;                                 /**< @brief Static storage for Mutex for synchronization of MQTT Context. */
     SemaphoreHandle_t contextMutex;                                        /**< @brief Mutex for synchronization of network buffer as the same buffer can be used my multiple applications. */
     uint8_t buffer[ NETWORK_BUFFER_SIZE ];                                 /**< @brief Network Buffer used to send packets on the network. This will be used by MQTT context defined above. */
-    NetworkContext_t networkContext;                                       /**< @brief Network Context used to send packets on the network. This will be used by MQTT context defined above. */
```
Should this really be removed? Shouldn't each mapping have both a context and params?

DanielYEHsieh replied:
Having both networkContext and the MQTTTransportParams seems a bit confusing, following other c files pattern I would recommend make networkContext declared in iot_mqtt_api.c

ravibhagavandas commented:
@DanielYEHsieh 
Could you update this unit test file as well: https://github.com/aws/amazon-freertos/blob/master/libraries/c_sdk/standard/ble/utest/ble_mqtt_transport/iot_ble_mqtt_transport_utest.c#L56
I think the cmock tests are failing because of that.

I agree that the connection timeout is unrelated. It seems to be a lab issue.

Commit:
avoid race conditiion issue for networkContext usnage in iot_mqtt_api.c

Commit:
1. remove coreHTTP check from mqtt CMake file
2. Apply network context change to iot_ble_mqtt_transport_utest.c

DanielYEHsieh commented:
@ravibhagavandas Thanks for pointing that out, fixed in 6e29b18

"Could you update this unit test file as well: https://github.com/aws/amazon-freertos/blob/master/libraries/c_sdk/standard/ble/utest/ble_mqtt_transport/iot_ble_mqtt_transport_utest.c#L56
I think the cmock tests are failing because of that.

I agree that the connection timeout is unrelated. It seems to be a lab issue."

Commit:
run uncrustify for iot_ble_mqtt_transport_utest.c

DanielYEHsieh commented:
BLE demos passed locally(CI is not working, tested with esp32 locally):
CONFIG_SHADOW_BLE_TRANSPORT_DEMO_ENABLED passed
CONFIG_MQTT_BLE_TRANSPORT_DEMO_ENABLED pass
https://paste.amazon.com/show/dhs/1607975879
https://paste.amazon.com/show/dhs/1607975352

DanielYEHsieh commented:
/bot run checks

Commit:
update more detailed description for NetworkContext

Commit:
provide more detailed description for NetworkContext in legacy MQTT

Commit:
Merge branch 'master' of github.com:aws/amazon-freertos into multiple-transport-interface

DanielYEHsieh commented:
/bot run checks

Commit:
run uncrustify

DanielYEHsieh commented:
/bot run checks

Commit:
fix build error after new demos are merged

yourslab commented:
/bot run checks

yourslab commented on demos/ble/mqtt_ble/mqtt_demo_ble_transport.c:
```diff
@@ -133,6 +133,20 @@
+ * It should contain a single pointer to the type of your desired transport.
+ * When using multiple transports in the same compilation unit, define this pointer as void *.
+ *
+ * @note Transport stacks are defined in FreeRTOS-Plus/Source/Application-Protocols/network_transport.
```
It's a different location for P4

DanielYEHsieh replied:
thanks, fixed in a57b636

Commit:
replace FreeRTOS-Plus/Source/Application-Protocols/network_transport. to
amazon-freertos/libraries/abstractions/transport/secure_sockets/transport_secure_sockets.h
and
amazon-freertos/libraries/c_sdk/standard/ble/include/iot_ble_mqtt_transport.h

replace amazon-freertos/libraries/c_sdk/standard/ble/include/iot_ble_mqtt_transport.h

DanielYEHsieh commented:
coreMQTT_Integration_BLE PASS!


---------STARTING TESTS---------
I (4816) BTDM_INIT: BT controller compile version [0fff744]
I (4816) system_api: Base MAC address is not set, read default base MAC address from BLK0 of EFUSE
GAP procedure initiated: advertise; disc_mode=2 adv_channel_map=0 own_addr_type=1 adv_filter_policy=0 adv_itvl_min=187 adv_itvl_max=375
GATT procedure initiated: notify; att_handle=94
GATT procedure initiated: notify; att_handle=94
GATT procedure initiated: notify; att_handle=94
GATT procedure initiated: notify; att_handle=94
GATT procedure initiated: notify; att_handle=94
GATT procedure initiated: notify; att_handle=94
GATT procedure initiated: notify; att_handle=94
GATT procedure initiated: notify; att_handle=94
TEST(coreMQTT_Integration_BLE, Subscribe_Publish_With_Qos_0) PASS
GATT procedure initiated: notify; att_handle=94
GATT procedure initiated: notify; att_handle=94
GATT procedure initiated: notify; att_handle=94
GATT procedure initiated: notify; att_handle=94
GATT procedure initiated: notify; att_handle=94
GATT procedure initiated: notify; att_handle=94
GATT procedure initiated: notify; att_handle=94
GATT procedure initiated: notify; att_handle=94
TEST(coreMQTT_Integration_BLE, Subscribe_Publish_With_Qos_1) PASS
GATT procedure initiated: notify; att_handle=94
GATT procedure initiated: notify; att_handle=94
GATT procedure initiated: notify; att_handle=94
TEST(coreMQTT_Integration_BLE, ProcessLoop_KeepAlive) PASS
GAP procedure initiated: terminate connection; conn_handle=0 hci_reason=19
GAP procedure initiated: advertise; disc_mode=2 adv_channel_map=0 own_addr_type=1 adv_filter_policy=0 adv_itvl_min=187 adv_itvl_max=375
GAP procedure initiated: stop advertising.

-----------------------
3 Tests 0 Failures 0 Ignored 
OK
-------ALL TESTS FINISHED-------

DanielYEHsieh commented:
unit test also passed for related changes

abstractions/transport/secure_sockets	
100.0% 100.0 %	164 / 164	100.0 %	8 / 8	100.0 %	90 / 90
c_sdk/standard/ble/src/services/data_transfer	
100.0% 100.0 %	392 / 392	100.0 %	28 / 28	100.0 %	170 / 170
c_sdk/standard/ble/src/services/mqtt_ble	
95.7% 95.7% 95.7 %	427 / 446	100.0 %	27 / 27	81.7 %	103 / 12
